File: java/largest_number.java
public class Solution {
    public String largestNumber(int[] nums) {
        List<String> numstr=new ArrayList<String>();
        for (int i:nums) {
            numstr.add(Integer.toString(i));
        }
        Collections.sort(numstr, new Comparator<String>(){
            @Override
            public int compare(String a, String b) {
                String s1=a+b;
                String s2=b+a;
                return s2.compareTo(s1);
            }
        });
        StringBuilder sb=new StringBuilder();
        
        for (String s:numstr) {
            sb.append(s);
        }
        
        while (sb.length()>1) {
            if (sb.charAt(0)=='0') {
                sb.deleteCharAt(0);
            } else {
                break;
            }
        }
        
        return sb.toString();
    }
}

File: cpp/largest-number.cpp
179. Largest Number  QuestionEditorial Solution  My Submissions
Total Accepted: 54991 Total Submissions: 268383 Difficulty: Medium
Given a list of non negative integers, arrange them such that they form the largest number.

For example, given [3, 30, 34, 5, 9], the largest formed number is 9534330.

Note: The result may be very large, so you need to return a string instead of an integer.

Credits:
Special thanks to @ts for adding this problem and creating all test cases.

Subscribe to see which companies asked this question

Hide Tags Sort

class Solution {
public:
    string largestNumber(vector<int>& nums) {
        string res;
        vector<string> v;
        for (auto n:nums) v.push_back(to_string(n));
        
        sort(v.begin(), v.end(), [](string& a, string& b) {return a+b>b+a; });
        for (auto s:v) res+=s;
        while (res.size()>1 && res[0]=='0') res.erase(res.begin());
        
        return res;
    }
};

